archiver/src/compress.rs

```diff
@@ -0,0 +1,215 @@
+use crate::{
+    error::Result,
+    utils::{ArchiveKind, Compression, PathExt},
+    Error,
+};
+use ignore::DirEntry;
+#[cfg(feature = "indicatif")]
+use indicatif::ProgressBar;
+use std::{
+    fs::File,
+    io::{self, Write},
+    path::PathBuf,
+};
+
+#[derive(Debug)]
+pub struct Archiver {
+    input: PathBuf,
+    output: PathBuf,
+    target_kind: Option<ArchiveKind>,
+
+    #[cfg(feature = "indicatif")]
+    progress_bar: Option<ProgressBar>,
```

Comment:
Indicatif support is not as simple as implementing `std::io::Write` in the archiver unfortunately because we need to have a complete picture of what gets compressed in order to have a determinate progress bar (which have better UX than indeterminate progress bars). I have some ideas to implement it more cleanly but I would rather have it tackled in a later PR if that suits you too